File: sdk/python/alp_sdk/formatter.py
"""ALP Formatter (v42.0.0 IDE Quality)"""

import os
from dataclasses import dataclass
from typing import List, Dict, Any, Optional


@dataclass
class FormatOptions:
    indent_size: int = 2
    preserve_comments: bool = True


@dataclass
class FormatResult:
    file: str
    formatted: bool


class AlpFormatter:
    def __init__(self, options: Optional[FormatOptions] = None):
        self.indent_size = (options.indent_size if options else 2)

    def format(self, content: str) -> str:
        lines = content.split('\n')
        formatted: List[str] = []
        in_directive = False
        object_indent = 0

        for line in lines:
            trimmed = line.strip()

            if trimmed == '':
                formatted.append('')
                continue

            if trimmed.startswith('!'):
                formatted.append(trimmed)
                in_directive = True
                continue

            if in_directive:
                formatted.append(trimmed)
                in_directive = False
                continue

            if trimmed.startswith('@'):
                formatted.append(trimmed)
                object_indent = 1
                continue

            if trimmed.startswith('[') or trimmed.startswith('<-'):
                formatted.append(self._indent(object_indent + 1) + trimmed)
                continue

            if ':' in trimmed:
                key, *rest = trimmed.split(':')
                value = ':'.join(rest).strip()
                if value == '':
                    formatted.append(self._indent(object_indent + 1) + trimmed)
                elif value.startswith('[') or value.startswith('<-') or value.startswith('@'):
                    formatted.append(self._indent(object_indent + 1) + trimmed)
                else:
                    formatted.append(self._indent(object_indent + 1) + f'{key}: {value}')
                continue

            formatted.append(self._indent(object_indent + 1) + trimmed)

        return '\n'.join(formatted)

    def format_file(self, file_path: str) -> str:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        return self.format(content)

    def format_workspace(self, alp_dir: str) -> List[FormatResult]:
        results: List[FormatResult] = []
        if not os.path.exists(alp_dir):
            return results

        def walk(directory: str):
            for entry in os.listdir(directory):
                full_path = os.path.join(directory, entry)
                if os.path.isdir(full_path):
                    walk(full_path)
                elif entry.endswith('.alp'):
                    original = self.format_file(full_path)
                    formatted = self.format(original)
                    changed = original != formatted
                    if changed:
                        with open(full_path, 'w', encoding='utf-8') as f:
                            f.write(formatted)
                    results.append(FormatResult(file=full_path, formatted=changed))

        walk(alp_dir)
        return results

    def _indent(self, level: int) -> str:
        return ' ' * (self.indent_size * level)

File: parser/src/formatter.ts
import * as fs from 'fs';
import * as path from 'path';

export interface FormatOptions {
  indentSize?: number;
  preserveComments?: boolean;
}

export class AlpFormatter {
  private indentSize: number;

  constructor(options: FormatOptions = {}) {
    this.indentSize = options.indentSize ?? 2;
  }

  public formatFile(filePath: string): string {
    const content = fs.readFileSync(filePath, 'utf8');
    return this.format(content);
  }

  public format(content: string): string {
    const lines = content.split('\n');
    const formatted: string[] = [];
    let inDirective = false;
    let objectIndent = 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const trimmed = line.trim();

      if (trimmed === '') {
        formatted.push('');
        continue;
      }

      if (trimmed.startsWith('!')) {
        formatted.push(trimmed);
        inDirective = true;
        continue;
      }

      if (inDirective) {
        formatted.push(trimmed);
        inDirective = false;
        continue;
      }

      if (trimmed.startsWith('@')) {
        formatted.push(trimmed);
        objectIndent = 1;
        continue;
      }

      if (trimmed.startsWith('[') || trimmed.startsWith('<-')) {
        formatted.push(this.indent(objectIndent + 1) + trimmed);
        continue;
      }

      if (trimmed.includes(':')) {
        const [key, ...rest] = trimmed.split(':');
        const value = rest.join(':').trim();
        if (value === '') {
          formatted.push(this.indent(objectIndent + 1) + trimmed);
        } else if (value.startsWith('[') || value.startsWith('<-')) {
          formatted.push(this.indent(objectIndent + 1) + trimmed);
        } else if (value.startsWith('@')) {
          formatted.push(this.indent(objectIndent + 1) + trimmed);
        } else {
          formatted.push(this.indent(objectIndent + 1) + `${key}: ${value}`);
        }
        continue;
      }

      formatted.push(this.indent(objectIndent + 1) + trimmed);
    }

    return formatted.join('\n');
  }

  public formatWorkspace(alpDir: string): { file: string; formatted: boolean }[] {
    const results: { file: string; formatted: boolean }[] = [];
    if (!fs.existsSync(alpDir)) return results;

    const walk = (dir: string) => {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
          walk(fullPath);
        } else if (entry.name.endsWith('.alp')) {
          const original = fs.readFileSync(fullPath, 'utf8');
          const formatted = this.format(original);
          const changed = original !== formatted;
          if (changed) {
            fs.writeFileSync(fullPath, formatted, 'utf8');
          }
          results.push({ file: fullPath, formatted: changed });
        }
      }
    };

    walk(alpDir);
    return results;
  }

  private indent(level: number): string {
    return ' '.repeat(this.indentSize * level);
  }
}
